http updates

hdonnay opened the pull request:
This patchset ends up making a few interrelated changes:

- Update the internal middleware to understand the `http.ResponseController` contract.
- Make use of the `ResponseController` pattern.
- Change the `apiError` helper to be impossible to use incorrectly from an `http.Handler`.
- Update the DiscoveryHandler to use the new style.
- Simplify the `httptransport.New` usage by not returning a custom type.

Supersedes and closes #1719.

hdonnay committed:
httptransport: exit goroutine in error helper

This change makes the `apiError` helper slightly tricksy in return for
making it impossible to misuse. Callers don't have to return after
calling `apiError;` instead the call just never returns.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: update DiscoveryHandler to new style

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: rework constructor

This makes the `New` function return a `http.Handler` instead of a
custom Server type.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: lints

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
compress: update compression middleware

This updates the compression middleware to use the 1.20
ResponseController scheme, zstd compression, and adds tests.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
webhook: move+update debug server

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httputil: add response recorder

This implements a new-style response recording `http.ResponseWriter`.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: re-instrument handlers with new primitives

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

hdonnay committed:
httptransport: use compression middleware

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

crozzy commented on middleware/compress/handler.go:
```diff
@@ -152,80 +153,77 @@ func (c *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		e1 = nil
 	}
-	c.next.ServeHTTP(nw, r)
+	return errors.Join(e0, e1)
```
Thoughts on calling these something more descriptive?

crozzy commented on httptransport/error.go:
```diff
@@ -52,7 +56,14 @@ func apiError(ctx context.Context, w http.ResponseWriter, code int, f string, v
+			Err(err).
+			Msg("unable to flush http response")
 	}
+	runtime.Goexit()
```
I'm a little on the fence about this, on the one hand I like the fact that it's terminal and it is always the last function called before the return anyway so why not incorporate the "return", however, the `_v1.go` files seem slightly harder to follow

```go
		report, ok, err := h.srv.IndexReport(ctx, d)
		if !ok {
			apiError(ctx, w, http.StatusNotFound, "index report not found")
		}
		if err != nil {
			apiError(ctx, w, http.StatusInternalServerError, "could not retrieve index report: %v", err)
		}
                // It appears like we'd get to here even if !ok and err != nil, but that isn't the case
		w.Header().Add("etag", validator)
```

hdonnay replied:
I suppose, but it's documented pretty well, IMO? I can add an additional callout in the docs if that would help.

crozzy commented on httptransport/server.go:
```diff
@@ -40,204 +40,96 @@ const (
-		e = t.configureComboMode(ctx)
-		if e != nil {
-			return nil, e
+	case config.IndexerMode, config.ComboMode:
```
So much cleaner but maybe we don't want a switch here and it'd be clearer with common or garden `if`s

```go
	if conf.Mode == config.IndexerMode || conf.Mode == config.ComboMode {
		if indexer == nil {
			return nil, fmt.Errorf("mode %q requires an indexer service", conf.Mode)
		}
		prefix := indexerRoot + apiRoot

		v1, err := NewIndexerV1(ctx, prefix, indexer, traceOpt)
		if err != nil {
			return nil, fmt.Errorf("indexer configuration: %w", err)
		}
		var sem *semaphore.Weighted
		if ct := conf.Indexer.IndexReportRequestConcurrency; ct > 0 {
			sem = semaphore.NewWeighted(int64(ct))
		}
		rl := &limitHandler{
			Check: func(r *http.Request) (*semaphore.Weighted, string) {
				if r.Method != http.MethodPost && r.URL.Path != IndexAPIPath {
					return nil, ""
				}
				// Nil if the relevant config option isn't set.
				return sem, IndexAPIPath
			},
			Next: v1,
		}
		mux.Handle(prefix, rl)
	}
	if conf.Mode == config.MatcherMode || conf.Mode == config.ComboMode {
		if indexer == nil || matcher == nil {
			return nil, fmt.Errorf("mode %q requires both an indexer service and a matcher service", conf.Mode)
		}
		prefix := matcherRoot + apiRoot
		v1 := NewMatcherV1(ctx, prefix, matcher, indexer, time.Duration(conf.Matcher.CacheAge), traceOpt)
		mux.Handle(prefix, v1)
	}
	if conf.Mode == config.NotifierMode || conf.Mode == config.ComboMode {
		if conf.Mode == config.ComboMode && notifier == nil {
			zlog.Debug(ctx).Msg("skipping unconfigured notifier")
			break // Skip
		}
		if notifier == nil {
			return nil, fmt.Errorf("mode %q requires a notifier service", conf.Mode)
		}
		prefix := notifierRoot + apiRoot
		v1, err := NewNotificationV1(ctx, prefix, notifier, traceOpt)
		if err != nil {
			return nil, fmt.Errorf("notifier configuration: %w", err)
		}
		mux.Handle(prefix, v1)
	}
```
Up to you

crozzy reviewed (commented):
The meat of this looks great, the couple of comments are pretty much style nits

hdonnay committed:
httptransport: add client-close detection

If an HTTP handler gets into the error handler, it now checks if the
request context is still valid. If not, the handler does not attempt to
write any response header or bodies and reports the non-standard 499
status code.

Signed-off-by: Hank Donnay <hdonnay@redhat.com>

crozzy reviewed (approved):
LGTM
